Multi-Room Support for Greenwave Reality

dfiel opened the pull request:
## Description:

Added support for gateways with multiple rooms already set up. No longer errors when more than 1 room is defined.


**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#4005

## Example entry for `configuration.yaml` (if applicable):
```yaml
light:
  - platform: greenwave
    host: 192.168.1.97
    version: 3
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

dfiel committed:
Multi-Room Support for Greenwave Reality

dfiel committed:
PEP8 Line Too Long

dfiel committed:
Update requirements_all.txt

tinloaf commented:
Looking at the existing code, it looks like the update logic is not optimal. Am I right assuming that every light will hit the XML API on its update, retrieving all information? I.e., the API will always be called multiple times? I think it would be way better to cache that XML document together with a time stamp somewhere (in `hass.data`?) and re-use it if the time stamp is not too old. One would have to force an update after service calls which change anything.

I'm not saying that this PR shouldn't be merged without this improvement, I'm just saying that if you feel up to the task… :wink:

dfiel commented:
@tinloaf Yes, your assumption is correct. This is modeled after the android app, but I do have plans to create an update function that updates all bulbs on one XML call. Hopefully in the next update.

dfiel commented:
Bump? I was hoping this would make it into the .61 release

pvizeli commented on homeassistant/components/light/greenwave.py:
```diff
@@ -104,9 +110,10 @@ def update(self):
+                    self._state = int(device['state'])
+                    self._brightness = greenwave.hass_brightness(device)
+                    self._online = greenwave.check_online(device)
+                    self._name = device['name']
```
That is a realy bad runtime. Make a data object like this: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/sensor/yweather.py#L174-L190

On this data object you can map it to a usefuly data format and access without loop.

You code actual:
- Hit the device with 10x bulbs it call the gateway 10x times
- Loop over 1000x

amelchio commented:
@dfiel an inefficient platform is frowned upon because it can slow down all of Home Assistant. What an Android app does is actually irrelevant in this regard.

However, this PR does not introduce the inefficiency but it rather fixes an error (that is even being reported now). So how about we merge this PR if @dfiel promises that the next PR will be to implement the shared state object?

dfiel commented:
Already working on it :)

dfiel committed:
Shared State Object

Shared State Object added, and implemented new function of greenwavereality to abstract complex runtimes.

dfiel commented:
@pvizeli Does this solve the issue? I believe it does.

pvizeli commented:
Yes

pvizeli committed:
Merge branch 'dev' into dev

pvizeli committed:
Update requirements_all.txt

pvizeli commented:
Please fix lint and we can merge it

dfiel committed:
Lint issues

dfiel committed:
Rewrite method to not trigger "Method could be a function"

dfiel commented:
@pvizeli Lint issues fixed

dfiel committed:
Remove unnecessary arguments for update call
